bench/macro/lsm-tree-bench-wp8.hs

```diff
@@ -347,29 +369,269 @@ doRun' gopts opts = do
+    LSM.withSession hasFS hasBlockIO (FS.mkFsPath []) $ \session -> do
+        -- open snapshot
+        -- In checking mode we start with an empty table, since our pure
+        -- reference version starts with empty (as it's not practical or
+        -- necessary for testing to load the whole snapshot).
+        tbl <- if opts.check
+                 then LSM.new  @IO @K @V @B session LSM.defaultTableConfig
+                 else LSM.open @IO @K @V @B session name
+
+        -- In checking mode, compare each output against a pure reference.
+        checkvar <- newIORef $ pureReference
+                                 gopts.initialSize opts.batchSize
+                                 opts.batchCount opts.seed
+        let check | not opts.check = \_ _ -> return ()
+                  | otherwise = \b y -> do
+              (x:xs) <- readIORef checkvar
+              unless (x == y) $
+                fail $ "lookup result mismatch in batch " ++ show b
+              writeIORef checkvar xs
+
+        let benchmarkIterations
+              | opts.pipelined = pipelinedIterations
+              | otherwise      = sequentialIterations
+        time <- timed_ $
+          benchmarkIterations
+            check
+            gopts.initialSize
+            opts.batchSize
+            opts.batchCount
             opts.seed
+            tbl
+
+        printf "Proper run (excluding snapshot open): %.03f sec\n" time
+        -- TODO: collect more statistic, save them in dry-run,
+        -- TODO: make the results human comprehensible.
 
-    withSession hasFS hasBlockIO (FS.mkFsPath []) $ \session -> do
-        -- open snapshot
-        tbl <- LSM.open @IO @K @V @B session name
 
-        void $ forFoldM_ initGen [ 0 .. opts.batchCount - 1 ] $ \b g -> do
-            let lookups :: [Word64]
-                inserts :: [Word64]
-                (!nextG, lookups, inserts) = generateBatch gopts.initialSize opts.batchSize g b
+-------------------------------------------------------------------------------
+-- sequential
+-------------------------------------------------------------------------------
+
+type LookupResults = V.Vector (K, LSM.LookupResult V ())
+
+{-# INLINE sequentialIteration #-}
+sequentialIteration :: (Int -> LookupResults -> IO ())
+                    -> Int
+                    -> Int
+                    -> LSM.TableHandle IO K V B
+                    -> Int
+                    -> MCG.MCG
+                    -> IO MCG.MCG
+sequentialIteration output !initialSize !batchSize !tbl !b !g = do
+    let (!g', ls, is) = generateBatch initialSize batchSize g b
+
+    -- lookups
+    results <- LSM.lookups ls tbl
+    output b (V.zip ls (fmap (fmap (const ())) results))
+
+    -- deletes and inserts
+    LSM.updates is tbl
+
+    -- continue to the next batch
+    return g'
+
+sequentialIterations :: (Int -> LookupResults -> IO ())
+                     -> Int -> Int -> Int -> Word64
```

Comment:
The types here make it easy to mix up arguments. A simple way of avoiding that would be to pass the whole `RunOpts`.

Reply:
I tend not to like passing UI things (CLI types) deep into the logic, but yes some structure would be sensible.